src/activity_actor.cpp

```diff
@@ -2686,6 +2692,473 @@ std::unique_ptr<activity_actor> disassemble_activity_actor::deserialize( JsonIn
     return actor.clone();
 }
 
+void read_activity_actor::start( player_activity &act, Character &who )
+{
+    if( !book->is_book() ) {
+        debugmsg( "ACT_READ on a non-book item" );
+        act.set_to_null();
+        return;
+    }
+
+    // book item_location must be of type character
+    // or else there will be item_location errors while loading
+    if( book.where() != item_location::type::character ) {
+        // character should have book before starting this activity
+        if( who.has_item( *book ) ) {
+            book = item_location( who, book.get_item() );
+        } else {
+            debugmsg( "ACT_READ character '%s' does not have '%s' book", who.get_name(), book->tname() );
+            act.set_to_null();
+            return;
+        }
+    }
+
+    bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
+             book_type::martial_art : book_type::normal;
+
+    // push copy of book for focus calculation
+    // avatar::update_mental_focus and avatar::calc_focus_equilibrium
+    act.targets.push_back( book );
+
+    act.moves_total = moves_total;
+    act.moves_left = moves_total;
+}
+
+void read_activity_actor::do_turn( player_activity &act, Character &who )
+{
+    if( who.fine_detail_vision_mod() > 4 ) {
+        //It got too dark during the process of reading, bail out.
+        act.set_to_null();
+        who.add_msg_if_player( m_bad, _( "It's too dark to read!" ) );
+        return;
+    }
+
+    if( !bktype.has_value() ) {
+        bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
+                 book_type::martial_art : book_type::normal;
+    }
+
+    if( who.is_avatar() ) {
```

Comment:
Needs to backport #46615 and solve merge conflict 😂
```suggestion
    if( !bktype.has_value() ) {
        bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
                 book_type::martial_art : book_type::normal;
    }

    if( who.is_avatar() ) {
        // the following check doesn't work for NPCs because it is counted for player's submap and z-level only
        if( who.fine_detail_vision_mod() > 4 ) {
            //It got too dark during the process of reading, bail out.
            act.set_to_null();
            who.add_msg_if_player( m_bad, _( "It's too dark to read!" ) );
            return;
        }
```